keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,354 @@
+https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-overview#pricing
+
+A user can ask for an internal Azure Load Balancer via a K8s Service definition that also has the annotation `service.beta.kubernetes.io/azure-load-balancer-internal: true`. There is not any limitation on the usage of mixed service protocols with internal LBs in the Azure CPI.
+
+Summary: Azure already has mixed protocol support for TCP and UDP, and it already affects the bills of the users. The implementation of this feature may require some work in Azure CPI.
+
+#### GCE
+
+The GCE CPI supports only TCP and UDP protocols in Services.
+
+GCE/GKE creates Network Load Balancers based on the K8s Services with type=LoadBalancer. In GCE there are "forwarding rules" that define how the incoming traffic shall be forwarded to the compute instances. A single forwarding rule can support either TCP or UDP but not both. In order to have both TCP and UDP forwarding rules we have to create separate forwarding rule instances for those. Two or more forwarding rules can share the same external IP if
+- the network load balancer type is External
+- the external IP address is not ephemeral but static
+
+There is a workaround in GCE, please see this comment from the original issue: https://github.com/kubernetes/kubernetes/issues/23880#issuecomment-269054735 If the external IP is static the user can create two Service instances, one for UDP and another for TCP and the user has to specify that static external IP address in those Service definitions in `loadBalancerIP`.
+
+HTTP protocol support: there is a different LB type in GCE for HTTP traffic: HTTP(S) LB. I.e. just like in the case of AWS it is not possible to have e.g. TCP+HTTP or UDP+HTTP behind the same LB/IP address.
+
+Forwarding rule pricing is per rule: there is a flat price up to 5 forwarding rule instances, on top of which every new forwarding rule has additional cost. 
+
+https://cloud.google.com/compute/network-pricing#forwarding_rules_charges
+
+[GCP forwarding_rules_charges](https://cloud.google.com/compute/network-pricing#forwarding_rules_charges) suggest that the same K8s Service definition would result in the creation of 2 forwarding rules in GCP. This has the same fixed price up to 5 forwarding rule instances, and each additional rule results in extra cost.
+
+A user can ask for an internal TCP/UDP Load Balancer via a K8s Service definition that also has the annotation `cloud.google.com/load-balancer-type: "Internal"`. Forwarding rules are also part of the GCE Internal TCP/UDP Load Balancer architecture, but in case of Internal TCP/UDP Load Balancer it is not supported to define different forwarding rules with different protocols for the same IP address. That is, for Services with type=LoadBalancer and with annotation `cloud.google.com/load-balancer-type: "Internal"` this feature would not be supported.
```

Comment:
How would a user find out this feature isn't supported for this particular case? It sounds like either searching logs or realizing the LB is not serving the forwarding rules as expected?

Reply:
This is an interesting point. The current GCP CPI implementation takes the first protocol value it finds in the Service definition and uses that for all ports in the Service. There is no check for violations of the "single protocol" rule, i.e. it assumes that the filter in the K8s API already blocked a mixed protocol setup. 
If this filter is removed by this feature then - according to my understanding - the GCP CPI will setup an internal load balancer with the same protocol on all forwarding rules, even if the trigger was a Service with mixed protocol setup. There is no error feedback to the user.

Reply:
We do not have arbitrary status, but we could add it.  `status.loadBalancer` could grow a field or two.  Or we could use events.

Reply:
Open question

Reply:
Indeed. Thank you! 

Which solution would you prefer: the new field(s) in the `status.loadBalancer`or  the events? 

I do not see them as contradictory, though. Events are moved to external troubleshooting systems pretty fast, so maybe it is worth to keep the actual state of the Service in-cluster too.

Reply:
I added now this idea to the "Proposed solution" chapter.

Reply:
I would love to see improvements to the LB status fields if possible.
At the very least, there should be some form of conditions to allow for better extensibility.

Reply:
Do you mean users can use different protocols on internal GCE LBs now?
I assume, documentation update is also coming? Here it still states that multiple forwarding rules for the same IP can have either UDP or TCP, but not both: https://cloud.google.com/load-balancing/docs/internal#multiple_forwarding_rule
And backend services also has similar limitation: https://cloud.google.com/load-balancing/docs/internal#backend-service

Or it is only my bad English parser :)

Reply:
> Do you mean users can use different protocols on internal GCE LBs now?

Yes :)

I'll ping docs people.  See https://cloud.google.com/kubernetes-engine/docs/how-to/internal-load-balancing#enabling_shared_ip

Reply:
Fixed.